tests/storage/databases/test_state_store.py

```diff
@@ -0,0 +1,109 @@
+from typing import Dict, List, Sequence, Tuple
+from unittest.mock import patch
+
+from twisted.internet.defer import Deferred, ensureDeferred
+
+from synapse.storage.state import StateFilter
+from synapse.types import MutableStateMap, StateMap
+
+from tests.unittest import HomeserverTestCase
+
+
+class StateGroupInflightCachingTestCase(HomeserverTestCase):
+    def setUp(self) -> None:
+        super(StateGroupInflightCachingTestCase, self).setUp()
+        # Patch out the `_get_state_groups_from_groups`.
+        # This is useful because it lets us pretend we have a slow database.
+        gsgfg_patch = patch(
+            "synapse.storage.databases.state.store.StateGroupDataStore._get_state_groups_from_groups",
+            self._fake_get_state_groups_from_groups,
+        )
+        gsgfg_patch.start()
+        self.addCleanup(gsgfg_patch.stop)
+        self.gsgfg_calls: List[
+            Tuple[Tuple[int, ...], StateFilter, Deferred[Dict[int, StateMap[str]]]]
+        ] = []
+
+    def prepare(self, reactor, clock, homeserver) -> None:
+        super(StateGroupInflightCachingTestCase, self).prepare(
+            reactor, clock, homeserver
+        )
+        self.state_storage = homeserver.get_storage().state
+        self.state_datastore = homeserver.get_datastores().state
+
+    def _fake_get_state_groups_from_groups(
+        self, groups: Sequence[int], state_filter: StateFilter
+    ) -> "Deferred[Dict[int, StateMap[str]]]":
+        print("hi", groups, state_filter)
+        d: Deferred[Dict[int, StateMap[str]]] = Deferred()
+        self.gsgfg_calls.append((tuple(groups), state_filter, d))
+        return d
+
+    def _complete_request_fake(
+        self,
+        groups: Tuple[int, ...],
+        state_filter: StateFilter,
+        d: "Deferred[Dict[int, StateMap[str]]]",
+    ) -> None:
+        """
+        Assemble a fake database response and complete the database request.
+        """
+
+        result: Dict[int, StateMap[str]] = {}
+
+        for group in groups:
+            group_result: MutableStateMap[str] = {}
+            result[group] = group_result
+
+            for state_type, state_keys in state_filter.types.items():
+                if state_keys is None:
+                    group_result[
+                        (state_type, "wild wombat")
+                    ] = f"{group} {state_type} wild wombat"
```

Comment:
Are the keys/values here important?

Reply:
It's been so long, but I think I was intending to write some more tests that check the behaviour of how subsets of state fit in, if the idea so far looked OK.

As it stands this could be really simplified, but instead the correct answer is probably for those tests to get written, if this path looks acceptable.